src/Orleans/Threading/GrainCancellationTokenSource.cs

```diff
@@ -0,0 +1,167 @@
+        /// <value>Whether cancellation has been requested for this <see
+        /// cref="Orleans.Threading.GrainCancellationTokenSource">CancellationTokenSource</see>.</value>
+        /// <remarks>
+        /// <para>
+        /// This property indicates whether cancellation has been requested for this token source, such as
+        /// due to a call to its
+        /// <see cref="Orleans.Threading.GrainCancellationTokenSource.Cancel()">Cancel</see> method.
+        /// </para>
+        /// <para>
+        /// If this property returns true, it only guarantees that cancellation has been requested. It does not
+        /// guarantee that every handler registered with the corresponding token has finished executing, nor
+        /// that cancellation requests have finished propagating to all registered handlers and remote targets. Additional
+        /// synchronization may be required, particularly in situations where related objects are being
+        /// canceled concurrently.
+        /// </para>
+        /// </remarks>
+        public bool IsCancellationRequested
+        {
+            get { return _cancellationTokenSource.IsCancellationRequested; }
+        }
+
+        /// <summary>
+        /// Communicates a request for cancellation.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The associated <see cref="T:Orleans.Threading.CancellationToken" /> will be
+        /// notified of the cancellation and will transition to a state where 
+        /// <see cref="Orleans.Threading.CancellationToken.IsCancellationRequested">IsCancellationRequested</see> returns true. 
+        /// Any callbacks or cancelable operations
+        /// registered with the <see cref="T:Orleans.Threading.CancellationToken"/>  will be executed.
+        /// </para>
+        /// <para>
+        /// Cancelable operations and callbacks registered with the token should not throw exceptions.
+        /// However, this overload of Cancel will aggregate any exceptions thrown into a <see cref="Orleans.AggregateException"/>,
+        /// such that one callback throwing an exception will not prevent other registered callbacks from being executed.
+        /// </para>
+        /// <para>
+        /// The <see cref="T:Orleans.Threading.ExecutionContext"/> that was captured when each callback was registered
+        /// will be reestablished when the callback is invoked.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="T:Orleans.AggregateException">An aggregate exception containing all the exceptions thrown
+        /// by the registered callbacks on the associated <see cref="T:Orleans.Threading.CancellationToken"/>.</exception>
+        /// <exception cref="T:Orleans.ObjectDisposedException">This <see
+        /// cref="T:Orleans.Threading.GrainCancellationTokenSource"/> has been disposed.</exception> 
+        public Task Cancel()
+        {
+            if (_cancellationTokenSource.IsCancellationRequested)
```

Comment:
In the case of transient exceptions, this will mean that exactly once semantics are *not* preserved. I think this check should be removed.